pkg/agent/agent.go

```diff
@@ -250,7 +252,7 @@ func startAPIServiceCache() {
 			return
 		}
 		// Start the full update after the first interval
-		go startDiscoveryCache(agent.instanceCacheLock)
+		go startDiscoveryCache(agent.instanceCacheLock, am)
 		log.Tracef("registered API cache update job: %s", id)
 	} else {
 		// Load cache from API initially. Following updates to cache will be done using watch events
```

Comment:
So, I changed the migration according to Jason's suggestions so that we didn't have an issue with some agents not having the agent resource. Now I realized that this isn't going to work for stream mode because we don't update the discovery cache with new resources when the agent starts up. This relates back to a defect I recently made where the resources won't get made if the sdk cache is outdated with the central environment, APIGOV-21724.

Reply:
@tsjohns9, with streaming mode when the persistent cache is turned on (by default turned off) we do not skip discovery rather we do not run the initialization for the cache by making the API call since we already have them loaded from the persistent store. To run the migration we should iterate over the cached api-service to perform the migration.

Also may be we should re-consider running the migration logic as part of the job which might have some race condition since we are adding more processing time with the migration logic, I understand it is one time as Migrate() method checks for x-agent-details subresource but if we have it running as part of job, other jobs might get kicked off while the discovery cache job is processing the migration and try to look for items in cache, so in my opinion the migration should be executed first and then we should start to initialize jobs as part of agent initialization, may be do a explicit call to Execute() method before starting the job like we do before starting the streaming job to update the cache in persistent cache is not available.